File: 0043-multiply-strings/0043-multiply-strings.cpp
class Solution {
public:
    string multiply(string num1, string num2) {
        int len1 = num1.size();
        int len2 = num2.size();
        vector<int> result(len1 + len2, 0);

        for (int i = len1 - 1; i >= 0; --i) {
            for (int j = len2 - 1; j >= 0; --j) {
                int mul = (num1[i] - '0') * (num2[j] - '0');
                int sum = mul + result[i + j + 1];
                result[i + j + 1] = sum % 10;
                result[i + j] += sum / 10;
            }
        }
        
        string res = "";
        for (int num : result) {
            res.push_back(num + '0');
        }

        auto it = res.begin();
        while (it != res.end() && *it == '0') {
            ++it;
        }

        return it == res.end() ? "0" : string(it, res.end());
    }
};

// class Solution {
// public:
//     string multiply(string num1, string num2) {
//         long n1 = 0, n2 = 0;
//         for (int i = 0; i < num1.length() ; i++) {
//             n1 = n1*10 + (num1[i] - '0');
//         }
//         for (int i = 0; i < num2.length() ; i++) {
//             n2 = n2*10 + (num2[i] - '0');
//         }
//         return to_string(n1*n2);
//     }
// };

File: LeetCode/0043-multiply-strings/0043-multiply-strings.py
class Solution:
    def multiply(self, num1: str, num2: str) -> str:
        m, n = len(num1), len(num2)
        pos = [0] * (m + n)
        
        for i in range(m - 1, -1, -1):
            for j in range(n - 1, -1, -1):
                mul = (ord(num1[i]) - ord('0')) * (ord(num2[j]) - ord('0'))
                p1, p2 = i + j, i + j + 1
                total = mul + pos[p2]
                pos[p1] += total // 10
                pos[p2] = total % 10
        
        result = ''.join(map(str, pos)).lstrip('0')
        return result if result else '0'